numpy/random/mtrand/distributions.c

```diff
@@ -719,26 +720,32 @@ double rk_wald(rk_state *state, double mean, double scale)
 
 long rk_zipf(rk_state *state, double a)
 {
-    double T, U, V;
-    long X;
+    double T, U, V, X;
     double am1, b;
 
     am1 = a - 1.0;
     b = pow(2.0, am1);
-    do
-    {
-        U = 1.0-rk_double(state);
+    T = 0.0;
+    while (1) {
+        U = 1.0 - rk_double(state);
         V = rk_double(state);
-        X = (long)floor(pow(U, -1.0/am1));
-        /* The real result may be above what can be represented in a signed
-         * long. It will get casted to -sys.maxint-1. Since this is
-         * a straightforward rejection algorithm, we can just reject this value
-         * in the rejection condition below. This function then models a Zipf
+        X = floor(pow(U, -1.0/am1));
+        /*
+         * The real result may be above what can be represented in a signed
+         * long. Since this is a straightforward rejection algorithm, we can
+         * just reject this value. This function then models a Zipf
          * distribution truncated to sys.maxint.
          */
+        if (X > LONG_MAX || X < 1.0) {
+            continue;
+        }
+
         T = pow(1.0 + 1.0/X, am1);
-    } while (((V*X*(T-1.0)/(b-1.0)) > (T/b)) || X < 1);
-    return X;
+        if (V*X*(T - 1.0)/(b - 1.0) <= T/b) {
```

Comment:
Does inverting this condition have any `nan`-related effect? What happens if `a == 1` before and after this patch?

Reply:
Or if you don't want to think about that, just leave it as before but replace the below `break` with `continue`, and `return (long) X` at the end of the loop.

Reply:
Yes, it works better :) A nan will compare false and the loop will repeat. I don't think any of the distributions were coded with nans in mind, but I don't think we will get any nans here. In any case, theoretically b > 1, and the function computing b is well behaved when b is near 1, so I don't think it is a problem in practice either.

One could rewrite the condition without divisions, and I think that would be safe, but that is a bigger modification.

Reply:
Of course, the library functions could be buggy, and `(b - 1.0)` negative, in which case getting rid of the divisions would be an improvement, but rejection based algorithms are always uncertain to the degree that the functions used change between libraries and architectures.

Reply:
My concern here is trapping the algorithm in an infinite loop on `nan`, where previously it would return.

In fact, `a == nan` would cause that behaviour, right?

Can you just flip the condition to what it was before:

```
if (V*X*(T - 1.0)/(b - 1.0) > T/b) {
    continue;
}
return (long) X;

Reply:
Hmm, seems it's already broken on `np.nan` in 1.12 anyway(#9829), so I guess this is out of scope

Reply:
The validation of `a` is in the calling routine, so that is where the `nan` check should be. I can fix that. I wouldn't be surprised if other routines can also fail with `nan`. Long loops for `a` slightly bigger than 1 indicates a faulty algorithm. Rejection should not happen that often in a good algorithm.